File: cole/day03_two-sum/two-sum.cpp
// beats 98.97%

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> cache;
        cache.reserve(nums.size());
        auto pos = 0;
        for (const auto &num : nums)
        {
            auto it = cache.find(target - num);
            if (it != cache.end())
            {
                return vector<int>{cache[target - num], pos};
            }
            cache[num] = pos++;
        }
        return vector<int>();
    }
};

File: cole/day03_two-sum/two-sum.py
# beats 89.71%

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = {}
        for i, num in enumerate(nums):
            if (target - num) in d:
                return (i, d[target - num])
            d[num] = i
